essence-of-live-coding/test/Cell.hs

```diff
@@ -23,10 +25,45 @@ import LiveCoding
 
 import qualified Cell.Util
 import qualified Cell.Monad.Trans
+import Control.Exception (ErrorCall)
+import Test.Framework.Providers.HUnit
+import Control.Exception.Base
+import Test.HUnit
 
 test = testGroup "Cell"
   [ testProperty "steps produces outputs"
     $ \(inputs :: [Int]) -> inputs === fst (runIdentity $ steps (id :: Cell Identity Int Int) inputs)
+  , testProperty "traverse' works as expected for Lists"
+    $ \(inputs :: [[Int]]) -> 
+      fmap (fmap (+10)) inputs === fst (runIdentity $ steps (traverse' (arr (+10) :: Cell Identity Int Int)) inputs)
+  , testProperty "traverse' works as expected for Maybes"
+    $ \(inputs :: [Maybe Int]) -> 
+      fmap (fmap (+10)) inputs === fst (runIdentity $ steps (traverse' (arr (+10) :: Cell Identity Int Int)) inputs)
+  , testCase "traverse' by itself does not force the entire list (ArrM)" 
+    $ testError "traverse' is too strict" id
+    $ head $ head $ fst $ runIdentity $ steps (traverse' (arr id) :: Cell Identity [Int] [Int]) [[1, error "Bang !"]]
+  , testCase "traverse' by itself does not force the entire list (Cell)"
+    $ testError "Traverse' is too strict" id
+    $ head $ head $ fst $ runIdentity $ steps (traverse' cellId :: Cell Identity [Int] [Int]) [[1, error "Bang !"]]
+  , testCase "traverse' of a Cell composed with another Cell does not force the entire list"
+    $ testError "Traverse' is too strict" id
+    $ head $ head $ fst $ runIdentity $ steps (traverse' cellId >>> cellId  :: Cell Identity [Int] [Int]) [[1, error "Bang !"]]
   , Cell.Util.test
   , Cell.Monad.Trans.test
-  ]
+  ]  
+
+testError :: String -> (Bool -> Bool) -> a -> IO ()
+testError message alterBoolValue value = do
+  let 
+    handler :: ErrorCall -> IO Bool
+    handler e = pure True
+  errored <- catch (pure $ seq value False) handler
```

Comment:
I'm confused. Will this be `True` when there was an exception? I believe not.

Reply:
Yes, I believe so. If there is no exception in value then `seq value False` returns `False`. If there is an exception then handler will run which will return `True`.

Reply:
But that means that the test _fails_ if there is an exception, right? For that use case, no extra function is needed. If an error is raised anywhere in a test, that counts as a test failure. What I wanted was a function where the test _succeeds_ if there is an exception in the value. That way you can test that a function has a certain strictness, by storing an exception somewhere in the place where you expect strictness, and making sure the whole thing explodes. But I think we don't have these places anymore, because `traverse'` is now lazy.

Maybe it's easiest to drop the tests with the errors, since we don't guarantee strictness or laziness anyways.

Reply:
> But that means that the test fails if there is an exception, right? 
yes.

> For that use case, no extra function is needed. If an error is raised anywhere in a test, that counts as a test failure.

Ah, I was missing that bit. Then I could change the tests to just a normal CellSimulation. Should I do that, or just drop the tests ?

>  What I wanted was a function where the test succeeds if there is an exception in the value. 

That is just a question of changing the  `errored` to `not errored` in the function `testError` , then it will fail the test if there is no error.

Reply:
> > For that use case, no extra function is needed. If an error is raised anywhere in a test, that counts as a test failure.
> 
> Ah, I was missing that bit. Then I could change the tests to just a normal CellSimulation. Should I do that, or just drop the tests ?
> 
> > What I wanted was a function where the test succeeds if there is an exception in the value.
> 
> That is just a question of changing the `errored` to `not errored` in the function `testError` , then it will fail the test if there is no error.

I think you can either drop the tests or change them to a `CellSimulation` if you want.